### PR TITLE
Add open telemetry to digitiser aggregator

### PR DESCRIPTION
**Pull request by Modularius**
Adds OpenTelemetry and some tracing to the aggregator component.

- otel_endpoint optional command line parameter added
- init_tracer function sets up OpenTelemetry if otel_endpoint is specified and runs tracing_subscriber::fmt::init() otherwise
- Extracts external trace from consumed Kafka headers if present and sets current trace as a child of it.
- Associates a trace with each Frame object, to which all relevant external traces are collected. (I'm not 100% happy with how fiddly this is - modifications are needed to the PartialFrame and AggregatedFrame functionality - it suggests there could be room to streamline the Spanned generic type in the future, but it works fine for the moment).

### CODE REVIEW

File: digitiser-aggregator/src/main.rs [hunk 1]
```diff
@@ -83,33 +90,44 @@ async fn main() {
...
 
 async fn on_message(
+    root_span: &Span,
+    args: &Cli,
     cache: &mut FrameCache<EventData>,
     producer: &FutureProducer,
-    output_topic: &str,
     msg: &BorrowedMessage<'_>,
 ) {
+    let span = trace_span!("Event Formation Message");
```
Comment (DanNixon) [inline digitiser-aggregator/src/main.rs:121]: This is a case that should just be done with a function annotation, not manually creating a span.

File: digitiser-aggregator/src/spanned_frame.rs [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+    fn new(ttl: Duration, metadata: FrameMetadata) -> Self {
+        let span = trace_span!("Frame");
+        Spanned {
+            span,
+            value: PartialFrame::<D>::new(ttl, metadata),
+        }
+    }
+}
+
+pub(crate) struct SpannedAggregatedFrame<D> {
```
Comment (DanNixon) [inline digitiser-aggregator/src/spanned_frame.rs:23]: Why can't `AggregatedFrame` also use `Spanned`?

**Review by DanNixon — CHANGES_REQUESTED** [review]

File: digitiser-aggregator/src/spanned_frame.rs (hunk 2 above)
Comment (Modularius) [reply digitiser-aggregator/src/spanned_frame.rs:23]: Due to Rust rules on implementing foreign traits on foreign types. I'm going to redesign this in a subsequent PR however.

Comment (DanNixon) [reply digitiser-aggregator/src/spanned_frame.rs:23]: I don't see any reason not to have the trait impl next to the type definition in this case though. The defining crate is in this repo and the fact there is an additional trait on the type should not adversely affect any other crate.

Comment (Modularius) [reply digitiser-aggregator/src/spanned_frame.rs:23]: It's a restriction on foreign types *and* foreign traits, so as Spanned<T> is from another crate, it won't allow compilation.
That is
```rust
SpannedAggregatedFrame<D> = Spanned<AggregatedFrame<D>>;

impl<D: Debug> AggregatedFrameLike<D> for Spanned<AggregatedFrame<D>> {
...
}
```
is not allowed.

Comment (DanNixon) [reply digitiser-aggregator/src/spanned_frame.rs:23]: Ah, I see, it's the `impl From` that is the issue. If this is going to be redesigned then it is fine for now.

File: digitiser-aggregator/src/main.rs [hunk 3]
```diff
@@ -83,33 +90,42 @@ async fn main() {
...
             }
             _ = cache_poll_interval.tick() => {
-                cache_poll(&mut cache, &producer, &args.output_topic).await;
+                cache_poll(&args, &mut cache, &producer).await;
             }
         }
     }
 }
 
+#[tracing::instrument(skip_all, name = "Event Formation Message", level = "trace")]
```
Comment (DanNixon) [inline digitiser-aggregator/src/main.rs:114]:
```suggestion
#[tracing::instrument(skip_all, level = "trace")]
```

I'd prefer just sticking with function names, if that is not clear when looking at the traces then the function should be renamed.

**Pull request merged**